Mathlib/Algebra/Module/Presentation/Basic.lean

```diff
@@ -0,0 +1,284 @@
+by `solution : relations.Solutions M`, this is the associated linear equivalence
+`relations.Quotient ≃ₗ[A] M`. -/
+noncomputable def linearEquiv : relations.Quotient ≃ₗ[A] M := LinearEquiv.ofBijective _ h.bijective
+
+@[simp]
+lemma linearEquiv_apply (x : relations.Quotient) :
+    h.linearEquiv x = solution.fromQuotient x := rfl
+
+@[simp]
+lemma linearEquiv_symm_var (g : relations.G) :
+    h.linearEquiv.symm (solution.var g) = Submodule.Quotient.mk (Finsupp.single g 1) :=
+  h.linearEquiv.injective (by simp)
+
+variable {N : Type v'} [AddCommGroup N] [Module A N]
+
+/-- If `M` admits a presentation by generators and relations, and we have a solution of the
+same equations in a module `N`, then this is the canonical induced linear map `M →ₗ[A] N`. -/
+noncomputable def desc (s : relations.Solution N) : M →ₗ[A] N :=
+  s.fromQuotient.comp h.linearEquiv.symm.toLinearMap
+
+@[simp]
+lemma desc_var (s : relations.Solution N) (g : relations.G) :
+    h.desc s (solution.var g) = s.var g := by
+  dsimp [desc]
+  rw [linearEquiv_symm_var, fromQuotient_single]
+
+include h in
+lemma postcomp_injective {f f' : M →ₗ[A] N}
+    (h' : solution.postcomp f = solution.postcomp f') : f = f' := by
+  suffices f.comp solution.fromQuotient = f'.comp solution.fromQuotient by
+    ext x
+    obtain ⟨y, rfl⟩ := h.bijective.2 x
+    exact DFunLike.congr_fun this y
+  ext g
+  simpa using congr_fun (congr_arg Solution.var h') g
+
+/-- If `M` admits a presentation by generators and relations, then
+linear maps from `M` can be (naturally) identified to the solutions of
+certain linear equations. -/
+@[simps]
+noncomputable def linearMapEquiv : (M →ₗ[A] N) ≃ relations.Solution N where
+  toFun f := solution.postcomp f
+  invFun s := h.desc s
+  left_inv f := h.postcomp_injective (by aesop)
+  right_inv s := by aesop
+
+end IsPresentation
+
+variable (relations)
+
+/-- Given `relations : Relations A`, this is the obvious solution to `relations`
+in the quotient `relations.Quotient`. -/
+noncomputable def ofQuotient : relations.Solution relations.Quotient :=
+  ofπ relations.toQuotient (by simp)
+
+@[simp]
+lemma ofQuotient_π : (ofQuotient relations).π = Submodule.mkQ _ := ofπ_π _ _
+
+@[simp]
+lemma ofQuotient_fromQuotient : (ofQuotient relations).fromQuotient = .id := by aesop
+
+lemma ofQuotient_isPresentation : (ofQuotient relations).IsPresentation where
+  bijective := by
+    simpa only [ofQuotient_fromQuotient, LinearMap.id_coe] using Function.bijective_id
+
+end Solution
+
+end Relations
+
+variable {A} (M : Type v) [AddCommGroup M] [Module A M] (relations : Relations A)
+
+/-- Given an `A`-module `M`, a term in this type is a presentation by `M` by
+generators and relations, where the shapes of the relations are given by
+a certain `relations : Relations A`. -/
+structure Presentation where
+  /-- a solution to the given equations -/
+  solution : relations.Solution M
+  isPresentation : solution.IsPresentation
```

Comment:
My first thought would be to bundle the `relations : Relations A` as well. What are the scenarios where we want a presentation with prescribed relations?

Reply:
Yes, I think you are right! I am updating the code.